designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,420 @@
+        engine.addPlugin(id, definition)
+      }
+    }
+  }
+}
+```
+
+</details>
+
+#### § The `executeOnFiles()` method
+
+This method returns an object that has two methods `then()` and `[Symbol.asyncIterator]()`, so we can use the returned object with `await` expression and `for-await-of` statement.
+
+- If you used the returned object with `for-await-of` statement, it iterates the lint result of each file in random order. This way yields each lint result immediately. Therefore you can print the results in streaming, or print progress state. ESLint may spend time to lint files (for example, ESLint needs about 20 seconds to lint our codebase), to print progress state will be useful.
+
+  ```js
+  const { ESLint } = require("eslint")
+  const eslint = new ESLint()
+
+  for await (const result of eslint.executeOnFiles(patterns)) {
+    print(result)
+  }
+  ```
+
+- If you used the returned object with `await` expression, it returns the array that contains all lint results in random order. This way waits for all lint results then returns it.
+
+  ```js
+  const { ESLint } = require("eslint")
+  const eslint = new ESLint()
+
+  const results = await eslint.executeOnFiles(patterns)
+  // Optionally you can sort the results.
+  // results.sort(ESLint.compareResultsByFilePath)
+  print(results)
+  ```
+
+Either way, we can support "linting in parallel", loading configs/plugins with `import()`, and other stuff that needs asynchronous logic.
+
+##### Move the `usedDeprecatedRules` property
+
+The returned object of `CLIEngine#executeOnFiles()` has the `usedDeprecatedRules` property that includes the deprecated rule IDs which the linting used. But the location doesn't fit the use with `for-await-of` statement. Therefore, this RFC moves the `usedDeprecatedRules` property to each lint result.
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+for await (const result of eslint.executeOnFiles(patterns)) {
+  console.log(result.usedDeprecatedRules)
+}
+```
+
+As a side-effect, formatters gets the capability to print the used deprecated rules. (!?)
+
+##### Fail-fast on parallel calling
+
+Because this method updates the cache file, if people call this method in parallel then it causes broken. To prevent the broken, it should throw an error if people called this method while the previous call is still running.
+
+##### Abort linting
+
+The iterator has optional [`return()` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return). The `for-of`/`for-await-of` syntax calls the `return()` method automatically if the execution escaped from the loop by `braek`, `return`, or `throw`. In short, the `return()` method will be called when aborted.
+
+Therefore, ESLint aborts linting when the `return()` method is called. This will mean that the `return()` method terminates all workers once we implement parallel linting.
```

Comment:
```suggestion
ESLint aborts all linting when the `return()` method is called. This will eg. terminate all workers if parallel linting is implemented.
```

+ add reference to that RFC